Jersey Validation Improvements

cmicali opened the pull request:
This PR is a few changes we have been using for a while to improve the built-in jersey parameter validation error messages provided by Dropwizard.

**Better Enum Validation Messages**
Currently dropwizard resources return a 404 if a request is made with an invalid value for a parameter of an `Enum` type. This PR includes `ValidatingEnumParamConverterProvider` which provides converters for all enums used in method parameters. The converter then provides better error messages like _"query param choice must be one of [OptionA, OptionB, OptionC]"_ when invalid values are sent.

**Parameter Names in Parse Failure Messages**
Currently when an invalid value is passed into a *Param parameter dropwizard returns an error message like _"Parameter is not a number."_.  This PR includes `AbstractParamConverterProvider` 
along with a slight enhancement to the *Param classes that returns messages that include the name of the parameter that failed validation, like _"query param user_id is not a number."_

**Hibernate ParameterNameProvider**
Small change to move most of parameter naming logic from `ConstraintMessage` to `JerseyParameterNameProvider`. This implements the standard `ParameterNameProvider` interface that Hibernate uses to resolve parameter names during validation.

**New DurationParam**
Include a new `DurationParam` that can parse String input in the format supported by dropwizard's existing `Duration` class.

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/params/DurationParam.java:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2016. Sense Labs, Inc. All Rights Reserved
```
Please remove copyright

cmicali replied:
D'oh, sorry - auto-added by intellij.

nickbabcock commented:
Awesome PR, did a quick glance and everything looks top notch.

I'm going to let this PR sit for a little bit as I mull over some of the implementation (like better enum validation messages).

nickbabcock commented on dropwizard-jersey/src/test/java/io/dropwizard/jersey/validation/ConstraintViolationExceptionMapperTest.java:
```diff
@@ -548,4 +549,60 @@ public void testInvalidFieldQueryParam() {
+            .queryParam("length", 50)
+            .request()
+            .get();
+        assertThat(response.getStatus()).isEqualTo(400);
```
Do you mind splitting up each of these assertions into a separate test case with a descriptive name like the other tests found in the file?

cmicali replied:
Yeah no problem - will do now.

cmicali commented:
Addressed comments on comments & tests and also fixed test failures related to `NonEmptyStringParam`.  The new converter for *Param types was not handling `NonEmptyStringParam` properly. I added support for it and removed `NonEmptyStringParamFeature` which was no longer needed.

cmicali commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/params/DurationParam.java:
```diff
@@ -0,0 +1,20 @@
+
+import io.dropwizard.util.Duration;
+
+public class DurationParam extends AbstractParam<Duration> {
```
This isn't directly related to the PR, but we use this often and seems like something Dropwizard should include by default.

nickbabcock replied:
Ideally, this may have been split into another PR, but that's ok here.

Can you write a test for this?

cmicali replied:
Done - updated description and added test.

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/params/AbstractParamConverterProvider.java:
```diff
@@ -0,0 +1,60 @@
+                // leaving Jersey to handle these parameters as it normally would.
+                return null;
+            }
+            return new ParamConverter<T>() {
```
Should this param converter be more like Jersey's default [`AbstractStringReader`](https://github.com/jersey/jersey/blob/927b4d0605aeb119d15e9623dab65c8aec2c8e20/core-server/src/main/java/org/glassfish/jersey/server/internal/inject/ParamConverters.java#L77-L112)?

cmicali replied:
Possibly - the one thing we might not want to change is the null handling. Jerey's provider will throw an `IllegalArgumentException` if the value is null.  My change does not, as this would be a breaking change with previous versions of dropwizard.

Otherwise, we could change the exception handling to match abstract string provider or we could inherit from AbstractStringReader and just modify the behavior we don't like. Thoughts?

nickbabcock replied:
Do you know how it will break backwards compatibility because doesn't the current classes get instantiated through [`StringConstructor`](https://github.com/jersey/jersey/blob/927b4d0605aeb119d15e9623dab65c8aec2c8e20/core-server/src/main/java/org/glassfish/jersey/server/internal/inject/ParamConverters.java#L118-L137), which calls [`AbstractStringReader`](https://github.com/jersey/jersey/blob/927b4d0605aeb119d15e9623dab65c8aec2c8e20/core-server/src/main/java/org/glassfish/jersey/server/internal/inject/ParamConverters.java#L77-L112)

cmicali replied:
I think you're right, I'm getting confused. Just committed an update that makes it basically match `AbstractStringReader` and added a test for not provide vs empty parameter. All tests pass so I think this is OK.

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/JerseyParameterNameProvider.java:
```diff
@@ -0,0 +1,65 @@
+public class JerseyParameterNameProvider extends ReflectionParameterNameProvider {
+
+    @Override
+    public List<String> getParameterNames(Constructor<?> constructor) {
```
Does this method need to be included? Looks like you're just delegating to `super`.

cmicali replied:
No, will remove.

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/ValidatingEnumParamConverterProvider.java:
```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2016. Sense Labs, Inc. All Rights Reserved
```
Oh no, another copyright 😛

cmicali replied:
Removed!

nickbabcock replied:
Looks like the copyright came back 😉

cmicali replied:
Removed, again.  Intellij: amazing 99.999% of the time.

evnm commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/params/AbstractParamConverterProvider.java:
```diff
@@ -0,0 +1,60 @@
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Type;
+
+public class AbstractParamConverterProvider implements ParamConverterProvider {
```
Could you add a brief Javadoc to this class explaining what it does, please?

evnm commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/JerseyParameterNameProvider.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Optional;
+
+public class JerseyParameterNameProvider extends ReflectionParameterNameProvider {
+
```
Rm extra newline

evnm commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/JerseyParameterNameProvider.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.List;
+import java.util.Optional;
+
+public class JerseyParameterNameProvider extends ReflectionParameterNameProvider {
```
Ditto brief Javadoc

evnm commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/ValidatingEnumParamConverterProvider.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.Set;
+
+@Provider
+public class ValidatingEnumParamConverterProvider<C extends Enum<C>> implements ParamConverterProvider {
```
Ditto brief Javadoc

cmicali replied:
Thanks for the feedback - newline removed and javadoc added to those 3 classes.

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/ValidatingEnumParamConverterProvider.java:
```diff
@@ -0,0 +1,150 @@
+ * include the parameter name and a list of valid values.</p>
+ */
+@Provider
+public class ValidatingEnumParamConverterProvider<C extends Enum<C>> implements ParamConverterProvider {
```
[I was able to cut the implementation in half](https://gist.github.com/nickbabcock/d34c360e7af44be4f8ed142c14e143cc) and make it somewhat compatible with the [`FuzzyEnumModule`](https://github.com/dropwizard/dropwizard/blob/9e01f7ec99e47f24f26698dbde7379e51b4e3aad/dropwizard-jackson/src/main/java/io/dropwizard/jackson/FuzzyEnumModule.java) that we use for Jackson enum serialization. I think it could be a good idea if params and body enums were handled the same way:

If there isn't anything fundamentally wrong with my suggested foundation, I'd like to see it integrated and improved upon 😄

cmicali replied:
😮 this is much better. Facepalm for the scanning work when all you need is

```java
 if (!rawType.isEnum()) {
    return null;
}
```

I will update this PR with this - should I include 

```java
 final String text = CharMatcher.WHITESPACE
                    .removeFrom(jp.getText())
                    .replace('-', '_')
                    .replace('.', '_');
```

from [`FuzzyEnumModule`](https://github.com/dropwizard/dropwizard/blob/9e01f7ec99e47f24f26698dbde7379e51b4e3aad/dropwizard-jackson/src/main/java/io/dropwizard/jackson/FuzzyEnumModule.java) to make it behave exactly the same way?

cmicali replied:
Two other questions:

1. Should I rename the class FuzzyEnumParamConverterProvider to make it match [`FuzzyEnumModule`](https://github.com/dropwizard/dropwizard/blob/9e01f7ec99e47f24f26698dbde7379e51b4e3aad/dropwizard-jackson/src/main/java/io/dropwizard/jackson/FuzzyEnumModule.java) ?
2. Should we move the 'fuzzy' part out into `dropwizard-util`, maybe to an `Enums` class?

nickbabcock replied:
No problem -- the power of open source 🎉 

To answer your questions:

- Yes it is probably prudent to make the implementations exactly the same.
- So extracting the shared code into an `Enums` class in `dropwizard-util` would be a fine idea
- And yes a rename to `FuzzyEnumParamConverterProvider`, I'm assuming, will help discover-ability

cmicali replied:
Changes made and committed.

nickbabcock replied:
Looks like there was a change that made `readsEnumsUsingToStringWithDeserializationFeatureOff` test case fail

cmicali replied:
Forgot to add that change to the previous commit - commited/pushed.

evnm commented on dropwizard-util/src/main/java/io/dropwizard/util/Enums.java:
```diff
@@ -0,0 +1,46 @@
+            }
+        }
+
+        //In some cases there are certain enums that don't follow the same patter across an enterprise.  So this
```
* Single space after slashes
* s/patter/pattern/

cmicali replied:
👍

evnm commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/validation/FuzzyEnumParamConverterProvider.java:
```diff
@@ -0,0 +1,69 @@
+                    return (T) constant;
+                }
+
+                throw new WebApplicationException(getErrorResponse(String.format("%s must be one of [%s]", parameterName, Joiner.on(", ").join(constants))));
```
This line's pretty long. Let's break it up and move the `Joiner` to a static field:

```
final String errorMsg = String.format(
    "%s must be one of [%s]", parameterName, joiner.join(constants));
throw new WebApplicationException(getErrorResponse(errorMsg));
```

cmicali replied:
Good idea - done 👍

evnm commented:
Thanks, @cmicali!

nickbabcock commented:
Updated release notes in b1c563e2

Thanks for the huge effort, @cmicali 😄

cmicali commented:
Happy to contribute back, not huge at all. Thanks for all the great detailed review, the result is significantly better for it.
